DBZ-1452 Add the possibility to insert "op" and "ts_ms" field to the flattened record

daanroosen-DS opened the pull request:
https://issues.redhat.com/browse/DBZ-1452

We needed to add the operation and timestamp field from the payload struct to the record. Therefore, we adapted the ExtractNewRecordState SMT by adding two boolean configuration options (one for "op" and one for "ts_ms").  It could be a nice addition to include this permanently in the SMT.

The solution is based on pull-request #968, which added the ability of adding fields from the source struct. The addSourceFields and makeUpdatedSchema functions were broken down into smaller parts.
The original functionality is completely preserved. However, we had to add a prefix to the fields in order to prevent duplicate field names (ts_ms for example). As prefix, we chose the name of the struct the field comes from. 

Any suggestions are welcome!

gunnarmorling commented on debezium-core/src/test/java/io/debezium/transforms/ExtractNewRecordStateTest.java:
```diff
@@ -367,8 +382,8 @@ public void testAddSourceFields() {
             final SourceRecord unwrapped = transform.apply(createRecord);
-            assertThat(((Struct) unwrapped.value()).get("__lsn")).isEqualTo(1234);
-            assertThat(((Struct) unwrapped.value()).getString("__version")).isEqualTo("version!");
+            assertThat(((Struct) unwrapped.value()).get("source__lsn")).isEqualTo(1234);
```
How comes that this has changed? That'll be quite disruptive for existing users, and it's not something we should do. That said, it feels like we've cornered ourselves with the design of the existing options. In hindsight, it'd have been better to have something more generic like:

```
add.fields=op,ts_ms,source.txid,source.lsn
add.headers=op,...
```

But I can't see right now a good way for getting there without breaking existing consumers.

gunnarmorling replied:
So here's what I suggest:

* deprecate `add.source.fields` and `operation.header`
* add `add.fields` and `add.fields`; they take _simple_ names generally as values, e.g. txid, lsn, op etc.; column names will be __txid, __lsn, __op etc., as today. With the exception of `ts_ms` that's not ambiguous; in order to add source.ts_ms, the value source.ts_ms will have to be given (for the sake of regularity, that "qualified" form would work for all the other fields, too, but generally won't be needed to be used.). If a field is given in a qualified way, it's field name will be qualified, too, e.g. __source_ts_ms

Would that work?

daanroosen-DS replied:
Thanks for your feedback @gunnarmorling. I agree that it should be more generic. I will implement it in the way you propose, with `add.fields` and `add.headers`.

gunnarmorling commented:
@jpechane, quick heads-up for you on this one; I think it should go into 1.0.x, too.

daanroosen-DS commented:
@gunnarmorling, hereby the functionality as you proposed. This is what I have done:

- I added `add.fields` and `add.headers`
- Fields can be specified as `op,lsn` (result: `__op` and `__lsn`) as well as `op,source.lsn` (result: result: `__op` and `__source__lsn`) 
- `operation.header` and `add.source.fields` are deprecated (kept the functionality as it was)
- The new functionality has priority (e.g. if both `add.fields` and `add.source.fields` are specified, the latter will ignored)
- updated the documentation

In the documentation, I marked `operation.header` and `add.source.fields` as deprecated. Should it be better to remove it from the documentation? I wasn't sure about that.

gunnarmorling commented on debezium-core/src/main/java/io/debezium/transforms/ExtractNewRecordStateConfigDefinition.java:
```diff
@@ -109,6 +109,25 @@ public static DeleteHandling parse(String value, String defaultValue) {
             .withWidth(ConfigDef.Width.LONG)
             .withImportance(ConfigDef.Importance.LOW)
             .withDefault("")
-            .withDescription("Adds each field listed from the 'source' element of the payload, prefixed with __ "
```
I think this is a left-over from the original work, right? I'll revert that, as the behavior of `add.source.fields` remains unchanged as per our discussion and your update.

daanroosen-DS replied:
Indeed, this can be reverted. I looked over this, I guess

gunnarmorling replied:
No worries, I'm applying a few changes (mostly cosmetic) and push a commit for all of it to this PR in a bit.

gunnarmorling commented:
@daanroosen-DS, ok, pushed one more commit to this PR. It was a bit more work than originally anticipated, because I realized that the code was based on the assumption that `source` would be the only sub-struct. But as of recently, that's not the case, there also can be a `transaction` struct. So I've reworked the parts that deal with retrieval of values from these substructs, by means of a new class `FieldReference` which encapsulates all the related logic.

@Naros, @jpechane, your feedback on this would be welcomed, in particular on the public "API" changes, as we were discussing those some time ago. Thanks!

gunnarmorling commented:
Superseded by the rebased and squashed PR #1259. Thanks a lot, @daanroosen-DS!

daanroosen-DS commented:
@gunnarmorling You're welcome! Thanks for your improvements!
